### PR TITLE
Pr for #4465

### PR DESCRIPTION
**Pull request by bquantump**
https://dev.azure.com/azure-mgmt-ex/DotNET%20Management%20SDK/_boards/board/t/DotNET%20Management%20SDK%20Team/Stories/?workitem=4465

### CODE REVIEW

File: Azure.ResourceManager.Core/Adapters/ModelAdapters.cs [hunk 1]
```diff
@@ -17,6 +17,18 @@ protected TrackedResource(ResourceIdentifier id, Location location, T data)
...
+            if(object.ReferenceEquals(id, null)){
+                Id = null;
+            }
+            else{
+                Id = id;
+            }
+            Location = location;
+            Model = data;
+        }
+
```
Comment (bquantump) [inline Azure.ResourceManager.Core/Adapters/ModelAdapters.cs:31]: Why not simplify this to Id = object.ReferenceEquals(id, null)  ? null : id;?

This fails interestingly enough. 

The CIL ouput differs in that it puts one thing on the allocation stack and the set it at the end, making the runtime have to think whatever will be set will be a string, causing the unwanted implicit this to be call.


image

File: Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs [hunk 2]
```diff
@@ -60,7 +57,8 @@ public override int GetHashCode()
 
         public static implicit operator ResourceIdentifier(string other)
         {
-            return new ResourceIdentifier(other);
+            var val = other == null ? throw new ArgumentNullException(nameof(other)) : new ResourceIdentifier(other);
```
Comment (m-nash) [inline Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs:60]: for style reasons can we change this to an if throw otherwise assign?
```suggestion
            if(other == null)
                throw new ArgumentNullException(nameof(other));
            var val = new ResourceIdentifier(other);
```

File: Azure.ResourceManager.Core/Adapters/ModelAdapters.cs [hunk 3]
```diff
@@ -17,6 +17,20 @@ protected TrackedResource(ResourceIdentifier id, Location location, T data)
             Model = data;
         }
 
+        protected TrackedResource(string id, Location location, T data)
```
Comment (m-nash) [inline Azure.ResourceManager.Core/Adapters/ModelAdapters.cs:20]: need some context as to why this is needed.  Is this to avoid the automatic implicit call, allowing us to have null as the id for "request" objects?

**Review by m-nash — CHANGES_REQUESTED** [review]

File: Azure.ResourceManager.Core/Adapters/ModelAdapters.cs (hunk 3 above)
Comment (bquantump) [reply Azure.ResourceManager.Core/Adapters/ModelAdapters.cs:20]: Exactly, if null get's passed into the other constructor, the implicit this get's called. This allows us to properly avoid that and set null for request types

File: client/Program.cs [hunk 4]
```diff
@@ -10,7 +10,7 @@ static void Main(string[] args)
             Scenario scenario = null;
             try
             {
-                scenario = ScenarioFactory.GetScenario(Scenarios.All);
```
Comment (allenjzhang) [inline client/Program.cs:13]: nit: Revert this change

File: Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs [hunk 5]
```diff
@@ -21,10 +21,7 @@ public class ResourceIdentifier : IEquatable<ResourceIdentifier>, IEquatable<str
         public ResourceIdentifier(string id)
         {
             Id = id;
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs:23]: let's add null check/verification at any public method entry:               if (string.IsNullOrWhiteSpace(id)) throw new ArgumentNullException(nameof(id));

File: Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs [hunk 6]
```diff
@@ -143,6 +141,10 @@ public override string ToString()
         /// <param name="id">A properly formed resource identity</param>
         protected virtual void Parse(string id)
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs:142]: >protected virtual [](start = 8, length = 17)

private? Why protected virtual? is ResourceIdentifier to be sub-classed?

File: Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs [hunk 7]
```diff
@@ -143,6 +141,10 @@ public override string ToString()
...
         {
+            if (id == null)
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs:144]: if this is private, these checks should be at public method entry point.

File: Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs [hunk 8]
```diff
@@ -143,6 +141,10 @@ public override string ToString()
...
+            {
+                throw new ArgumentNullException(nameof(id));
+            }
             // Throw for null, empty, and string without the correct form
             if (string.IsNullOrWhiteSpace(id) || !id.Contains('/'))
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs:149]: I think we can use a simple regex to do validation:  /.*

File: Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs [hunk 9]
```diff
@@ -143,6 +141,10 @@ public override string ToString()
...
                 throw new ArgumentOutOfRangeException($"'{id}' is not a valid resource");
```
Comment (allenjzhang) [inline Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs:150]: Should be ArgumentException.

**Comment by allenjzhang** [comment]
>             if (!(KnownKeys.Subscription.Equals(parts[0], StringComparison.InvariantCultureIgnoreCase) ||

Added a discussion point on implementation in Sync meeting for 1/13.

---
Refers to: Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs:161 in 729d3c1. [](commit_id = 729d3c13bb7caad4ed9ce473f9aa867ccb651ba2, deletion_comment = False)

**Comment by allenjzhang** [comment]
>             public static string UntrackedSubResource => Guid.NewGuid().ToString();

I am not sure that untracked resources are all GUID.

---
Refers to: Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs:250 in 729d3c1. [](commit_id = 729d3c13bb7caad4ed9ce473f9aa867ccb651ba2, deletion_comment = False)

**Comment by allenjzhang** [comment]
>         public static class KnownKeys

private?

---
Refers to: Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs:238 in 729d3c1. [](commit_id = 729d3c13bb7caad4ed9ce473f9aa867ccb651ba2, deletion_comment = False)

File: Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs (hunk 5 above)
Comment (bquantump) [reply Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs:23]: Null check will be added when we decided on #118

File: Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs (hunk 9 above)
Comment (bquantump) [reply Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs:150]: Bring up at meeting in general to go over parsing ID

**Review by m-nash — CHANGES_REQUESTED** [review]
Can you add tests which now verify the original issue no longer exists?

**Review by m-nash — APPROVED** [review]

File: Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs (hunk 2 above)
Comment (bquantump) [reply Azure.ResourceManager.Core/Resources/ResourceIdentifier.cs]: I will add null checks with the outcome of #118
